NTupleAnalyzer_nanoAOD.py

```diff
@@ -1114,31 +1117,118 @@ def getMuonSF(_pt,_eta):
+    # https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2017#Momentum_Resolution
+    # https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2018#Momentum_Resolution
+    # Note: _p is muon 3-momentum--not transverse momentum (pt)
+    coeffsAllEta = [[0.0,0.0,0.0,0.0,0.0],[0.0,0.0,0.0,0.0,0.0],[0.0,0.0,0.0,0.0,0.0]]
+    coeffs = coeffsAllEta[0]
+
+    if _year == '2016': 
+        coeffsAllEta = [[0.0061, 0.0001, 0.000000110, 0.0000000000610, 0.0000000000000110],[0.0134, 0.00006, 0.0000000510, 0.0000000000310, 0.00000000000000510],[0.0151, 0.0001, 0.0000000410, 0.00000000000410, -0.0000000000000110]]
+
+    elif _year == '2017':
+        coeffsAllEta = [[0.0053, 0.0001, 0.000000110, 0.0000000000710, 0.0000000000000110],[0.0136, 0.00006, 0.0000000310, 0.00000000000110, 0.00000000000000310],[0.0170, 0.00008, 0.0000000310, 0.0000000000210, 0.0000000000000810]]
+
+    elif _year == '2018':
+        coeffsAllEta = [[0.0062, 0.0001, 0.000000110, 0.0000000000510, 0.00000000000000910],[0.0136, 0.00005, 0.0000000210, 0.00000000000510, 0.0],[0.0174, 0.00009, 0.00000000310, 0.0000000000210, 0.00000000000000510]]
+
+    if abs(_eta) <= 1.2 : coeffs = coeffsAllEta[0]
+    elif 1.2 < abs(_eta) <= 2.1 : coeffs = coeffsAllEta[1]
+    elif 2.1 < abs(_eta) <= 2.4 : coeffs = coeffsAllEta[2]
+
+    sigmaToReturn = coeffs[0] + coeffs[1]*_p - coeffs[2]*_p*_p + coeffs[3]*_p*_p*_p - coeffs[4]*_p*_p*_p*_p
+
+    return sigmaToReturn
+
+def SmearMuonCollections(_ptCollection, _etaCollection, _phiCollection, isSystematic):
+	# Following perscription for muon momentum smearing of 15% from resolution uncertainty
+    # https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2016#Momentum_Resolution
+	# https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2017#Momentum_Resolution
+    # https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2018#Momentum_Resolution
+
+	smearedPtCollection = _ptCollection
+	smearedEtaCollection = _etaCollection
+	smearedPhiCollection = _phiCollection
+
+	# loop through muons
+	for n in range(len(_ptCollection)):
+		if abs(_etaCollection[n]) <= 1.2 and not isSystematic: continue # no smearing if abs(eta) < 1.2
+		elif abs(_etaCollection[n]) > 1.2 or isSystematic:
+			smearConst = 0.57 # resolution smearing of 15% -> 0.57
+			if isSystematic: smearConst = 0.46 # systematics requires 10% smearing -> 0.46
+			#print 'pt:',_ptCollection[n]
+			#print 'eta:',_etaCollection[n]
+			#print 'phi:',_phiCollection[n]
+			smearedLorentz = TLorentzVector()
+			origLorentz = TLorentzVector()
+			origLorentz.SetPtEtaPhiM(_ptCollection[n], _etaCollection[n], _phiCollection[n], 0)
+			#print 'origLorentz is',origLorentz
+			# Smearing is performed on 3-momentum, convert from Pt, Eta, Phi to cartesian 4-momentum and back
+			origPx = origLorentz.Px()
+			origPy = origLorentz.Py()
+			origPz = origLorentz.Pz()
+
+			# Smear momenta here
+			smearedPx = origPx*(1 + tRand.Gaus(0.0, MERParametrization(origPx, _etaCollection[n])*smearConst))
```

Comment:
So this smears px,py,pz by different amounts. Shouldn't they all be smeared by the same amount? Did you get this code example from somewhere?

Reply:
I had to code this myself. You're right though, and it should be changed to smear each p component by the same amount. I'll fix that